be/src/vec/olap/block_aggregator.cpp

```diff
@@ -0,0 +1,234 @@
+          _return_columns(return_columns),
+          _null_set(null_set),
+          _batch_size(batch_size) {
+    _has_agg_data = false;
+    _do_aggregate = true;
+    _agg_ratio = config::block_aggregate_ratio;
+    _num_columns = _output_columns.size();
+    _num_key_columns = 0;
+
+    // todo(zeno) Make sure column prune done
+    std::sort(_sorted_output_columns.begin(), _sorted_output_columns.end());
+    _output_columns_loc.resize(_num_columns);
+    for (int i = 0; i < _num_columns; ++i) {
+        for (int j = 0; j < _num_columns; ++j) {
+            if (_sorted_output_columns[i] == output_columns[j]) {
+                _output_columns_loc[i] = j;
+                break;
+            }
+        }
+    }
+
+    for (auto index : _sorted_output_columns) {
+        if (_tablet_schema->column(index).is_key()) {
+            _num_key_columns++;
+        }
+    }
+
+    _current_row = 0;
+    _source_size = 0;
+    _eq_previous.reserve(_batch_size);
+    _agg_index.reserve(_batch_size);
+    _agg_range.reserve(_batch_size);
+
+    // init _key_comparator
+    for (int i = 0; i < _num_key_columns; ++i) {
+        _key_comparator.emplace_back(
+                _get_comparator(_tablet_schema->column(_sorted_output_columns[i])));
+    }
+
+    // init _column_aggregator
+    for (int i = 0; i < _num_key_columns; ++i) {
+        _column_aggregator.emplace_back(std::make_unique<KeyColumnAggregator>());
+    }
+
+    for (int i = _num_key_columns; i < _num_columns; ++i) {
+        bool is_nullable = (_null_set != nullptr &&
+                            _null_set->find(_sorted_output_columns[i]) != _null_set->end());
+        auto col = _tablet_schema->column(_sorted_output_columns[i]);
+        auto data_type = vectorized::DataTypeFactory::instance().create_data_type(col, is_nullable);
+        _column_aggregator.emplace_back(std::make_unique<ValueColumnAggregator>(col, data_type));
+    }
+
+    aggregate_reset();
+}
+
+CompareFunc BlockAggregator::_get_comparator(const TabletColumn& col) {
+    switch (col.type()) {
+    case OLAP_FIELD_TYPE_TINYINT:
+        return &compare_previous<ColumnInt8>;
+    case OLAP_FIELD_TYPE_SMALLINT:
+        return &compare_previous<ColumnInt16>;
+    case OLAP_FIELD_TYPE_INT:
+        return &compare_previous<ColumnInt32>;
+    case OLAP_FIELD_TYPE_BIGINT:
+        return &compare_previous<ColumnInt64>;
+    case OLAP_FIELD_TYPE_LARGEINT:
+        return &compare_previous<ColumnInt128>;
+    case OLAP_FIELD_TYPE_BOOL:
+        return &compare_previous<ColumnUInt8>;
+    case OLAP_FIELD_TYPE_FLOAT:
+        return &compare_previous<ColumnFloat32>;
+    case OLAP_FIELD_TYPE_DOUBLE:
+        return &compare_previous<ColumnFloat64>;
+    case OLAP_FIELD_TYPE_DECIMAL:
+        return &compare_previous<ColumnDecimal<Decimal128>>;
+    case OLAP_FIELD_TYPE_CHAR:
+    case OLAP_FIELD_TYPE_VARCHAR:
+    case OLAP_FIELD_TYPE_STRING:
+        return &compare_previous<ColumnString>;
+    case OLAP_FIELD_TYPE_DATE:
```

Comment:
OLAP_FIELD_TYPE_DATEV2,OLAP_FIELD_TYPE_DATETIMEV2,OLAP_FIELD_TYPE_DECIMAL32,OLAP_FIELD_TYPE_DECIMAL64,OLAP_FIELD_TYPE_DECIMAL128 also need to be processed

Reply:
done